### ISSUE TITLE
Defer JDK stream creation until necessary and merge Custom* streams with normal ones

### ISSUE THREAD

**Issue by amaembo** [amaembo/streamex#57]
A big refactoring effort could be made to defer creation of JDK stream until it's really necessary. This would allow to stick with spliterators for most of sources and quasi-intermediate operations and sometimes don't create Stream at all. For example `StreamEx.of(array).pairMap(blahblah).forEach(blahblah)`.

This also assumes merging `CustomStreamEx` and friends into their parents eliminating several classes (and actually reducing the library size a little bit).

**Comment by amaembo** [comment]
Benchmarking report for very simple pipeline creation (all times in us):

test | 0.5.3 | 0.5.4
---|---:|---:
flatOptional | 420,880 | 240,456
mapForEach | 0,066 | **0,073**
mapForEachInt | 0,062 | **0,073**
mapPairMapForEach | 0,138 | 0,120
mapPairMapForEachInt | 0,143 | 0,122
mapPairMapReduce | 0,162 | 0,163
mapPairMapReduceInt | 0,236 | 0,230
mapReduce | 0,061 | **0,068**
mapReduceInt | 0,067 | **0,070**
pairMapForEach | 0,083 | 0,047
pairMapForEachInt | 0,085 | 0,046
pairMapMapForEach | 0,117 | 0,103
pairMapMapForEachInt | 0,121 | 0,101
pairMapMapReduce | 0,114 | 0,106
pairMapMapReduceInt | 0,122 | 0,100
pairMapReduce | 0,092 | 0,073
pairMapReduceInt | 0,092 | 0,076
simpleForEach | 0,034 | 0,018
simpleForEachInt | 0,032 | 0,017
simpleReduce | 0,040 | 0,040
simpleReduceInt | 0,035 | 0,036
chain2 | 0,090 | **0,099**
chain3 | 0,134 | 0,135
chain4 | 0,164 | **0,167**
chain5 | 0,195 | **0,200**
chain6 | 0,224 | **0,241**

Very simple pipelines (2-3 elements) were created and executed here. The `chainX` test is just a chain of X `map` operations. New version has regress on some tests which involve normal stream API calls only (no quasi-intermediate operations). Adding quasi-intermediate ops usually makes things faster. Also connecting `forEach` directly to the source is faster as it does not create JDK stream at all. Due to this `flatOptional` test improved significantly (which involves `.flatMap(opt -> StreamEx.of(opt))`). Probably such fast paths should be implemented for other operations as well.

The `headTail` recursive functions imrpoved significantly:

test | 0.5.3 | error | 0.5.4 | error
---|---:|---:|---:|---:
htComplex | 7554,683 | 188,386 | 6883,196 | 140,810
htSimple | 1312,000 | 19,084 | 696,813 | 7,112
jdkComplex | 6002,417 | 273,612 | 6242,036 | 264,505
jdkSimple | 80,825 | 2,827 | 83,622 | 2,133

In conclusion this optimization is considered good.
